pkg/provider/github/github.go

```diff
@@ -68,137 +83,360 @@ func (w *GitHub) Create(ctx context.Context, obj *webhookv1.GitWatcher) (*webhoo
+	events := getEvents(obj)
+	hook, resp, err := client.Repositories.CreateHook(ctx, owner, repo, &github.Hook{
+		Events: events,
+		Config: map[string]interface{}{
+			"url":    getHookEndpoint(obj, obj.Spec.ReceiverURL),
+			"secret": obj.Status.Token,
 		},
+	})
+	if err != nil {
+		return obj, fmt.Errorf("failed to create hook for %s/%s, error: %v", owner, repo, err)
+	}
+	defer resp.Body.Close()
+
+	if resp != nil && resp.StatusCode != http.StatusCreated {
+		msg, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return obj, fmt.Errorf("failed to create hook, failed to read api response")
+		}
+		return obj, fmt.Errorf("failed to create hook for %s/%s, error: %v", owner, repo, msg)
 	}
 
-	hook, _, err := scmClient.Repositories.CreateHook(context.Background(), repoName, in)
-	if err != nil {
-		return obj, err
+	if hook != nil && hook.ID != nil {
+		obj.Status.HookID = strconv.Itoa(int(*hook.ID))
 	}
 
-	obj.Status.HookID = hook.ID
 	return obj, nil
 }
 
-func (w *GitHub) getClient(obj *webhookv1.GitWatcher) (*scm.Client, error) {
-	secret, err := w.GetSecret(secretName, obj)
+func getEvents(obj *webhookv1.GitWatcher) []string {
+	var events []string
+	if obj.Spec.Push {
+		events = append(events, "push")
+	}
+
+	if obj.Spec.PR {
+		events = append(events, "pull_request")
+	}
+
+	if obj.Spec.Tag {
+		events = append(events, "create")
+	}
+	return events
+}
+
+func (w *GitHub) getClient(ctx context.Context, obj *webhookv1.GitWatcher) (*github.Client, error) {
+	secretName := defaultSecretName
+	if obj.Spec.GithubWebhookToken != "" {
+		secretName = obj.Spec.GithubWebhookToken
+	}
+
+	secret, err := w.secretCache.Get(obj.Namespace, secretName)
 	if err != nil {
 		return nil, err
 	}
 
-	return newGithubClient(string(secret.Data["accessToken"]))
+	return newGithubClient(ctx, string(secret.Data["accessToken"])), nil
 }
 
-func (w *GitHub) HandleHook(gitCommits v1.GitWatcherCache, req *http.Request) (*webhookv1.GitWatcher, scm.Webhook, bool, int, error) {
+func (w *GitHub) HandleHook(ctx context.Context, req *http.Request) (int, error) {
 	receiverID := req.URL.Query().Get(utils.GitWebHookParam)
 	if receiverID == "" {
-		return nil, nil, false, 0, nil
+		return 0, nil
 	}
 
 	ns, name := kv.Split(receiverID, ":")
-	receiver, err := gitCommits.Get(ns, name)
+	gitwatcher, err := w.gitWatchers.Get(ns, name, metav1.GetOptions{})
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	if !gitwatcher.Spec.Enabled {
+		return http.StatusUnavailableForLegalReasons, errors.New("webhook receiver is disabled")
+	}
+
+	payload, err := github.ValidatePayload(req, []byte(gitwatcher.Status.Token))
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	event, err := github.ParseWebHook(github.WebHookType(req), payload)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	client, err := w.getClient(ctx, gitwatcher)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	w.client = client
```

Comment:
don't save, pass as args.  Also, does the client need to be closed?  what about http connection?